### DISCUSSION TITLE
Confusion Surrounding the JJ CLI - Seeking Clarification

### DISCUSSION THREAD

**Discussion by johnwheeler** [jj-vcs/jj#2691] [Q&A] [answered]
## Background

This issue arises from a discussion on Hacker News ([link](https://news.ycombinator.com/item?id=38590080#38593087)), where I was advised to post an example on the issue tracker.

### Steps Taken

1. Created and cloned a fresh repo from GitHub:

```bash
jwheeler@gregory Projects % jj git clone git@github.com:johnwheeler/jjtest1.git
```

2. Initial log check and file addition:

```bash
jwheeler@gregory Projects % cd jjtest1
jwheeler@gregory jjtest1 % echo a > file1
jwheeler@gregory jjtest1 % jj commit -m "added file1 with line a"
jwheeler@gregory jjtest1 % jj log
```

```
@  zlpppmms john@gilfoyle.local 2023-12-11 08:51:29.000 -08:00 35ceafe8
│  (empty) (no description set)
◉  sotnpuyn john@gilfoyle.local 2023-12-11 08:51:29.000 -08:00 78a03ac2
│  added file1 with line a
◉  zzzzzzzz root() 00000000
```

3. Adding another line to the file and committing:

```bash
jwheeler@gregory jjtest1 % echo b >> file1
jwheeler@gregory jjtest1 % jj commit -m "added line b to file1"
jwheeler@gregory jjtest1 % jj log
```

```
@  oovouqyn john@gilfoyle.local 2023-12-11 08:52:06.000 -08:00 dcd729c1
│  (empty) (no description set)
◉  zlpppmms john@gilfoyle.local 2023-12-11 08:52:06.000 -08:00 a11a89f8
│  added line b to file1
◉  sotnpuyn john@gilfoyle.local 2023-12-11 08:51:29.000 -08:00 78a03ac2
│  added file1 with line a
◉  zzzzzzzz root() 00000000
```

4. Pushing changes and re-cloning the repo:

```bash
jwheeler@gregory jjtest1 % jj git push -c @-
jwheeler@gregory Projects % rm -rf jjtest1
jwheeler@gregory Projects % jj git clone git@github.com:johnwheeler/jjtest1.git
jwheeler@gregory Projects % cd jjtest1
jwheeler@gregory jjtest1 % jj log
```

At this point, the log history still looks linear. I get the same output as above for jj log, which I’ll omit here.

5. Making another change, pushing it, and re-cloning:

```bash
jwheeler@gregory jjtest1 % echo a >> file2
jwheeler@gregory jjtest1 % jj commit -m "added file2 with line a"
jwheeler@gregory jjtest1 % jj git push -c @-
jwheeler@gregory Projects % rm -rf jjtest1
jwheeler@gregory Projects % jj git clone git@github.com:johnwheeler/jjtest1.git
jwheeler@gregory Projects % cd jjtest1
jwheeler@gregory jjtest1 % jj log
```

```
@  vvlzyqsx john@gilfoyle.local 2023-12-11 08:54:14.000 -08:00 1acd8dfe
│  (empty) (no description set)
│ ◉  rokrzznm john@gilfoyle.local 2023-12-11 08:53:34.000 -08:00 push-lrpmlyoprlnk 7ec59c8a
├─╯  added file2 with line a
◉  krzmzxtv john@gilfoyle.local 2023-12-11 08:52:06.000 -08:00 push-zlpppmmsqsyu a11a89f8
│  added line b to file1
◉  kppvnyqs john@gilfoyle.local 2023-12-11 08:51:29.000 -08:00 78a03ac2
│  added file1 with line a
◉  zzzzzzzz root() 00000000
```

At this point I become confused. I expect the history to be linear, but I see what appears to be a branch. 

6. Observing a non-linear history after merging a PR on GitHub and fetching:

```bash
jwheeler@gregory jjtest1 % jj git fetch
jwheeler@gregory jjtest1 % jj log
```

```
◉    yxynovlq jwheeler726@gmail.com 2023-12-11 09:01:42.000 -08:00 push-zlpppmmsqsyu 4b287014
├─╮  (empty) Merge pull request #1 from johnwheeler/push-lrpmlyoprlnk
│ ◉  rokrzznm john@gilfoyle.local 2023-12-11 08:53:34.000 -08:00 push-lrpmlyoprlnk 7ec59c8a
├─╯  added file2 with line a
│ @  vvlzyqsx john@gilfoyle.local 2023-12-11 08:54:14.000 -08:00 1acd8dfe
├─╯  (empty) (no description set)
◉  krzmzxtv john@gilfoyle.local 2023-12-11 08:52:06.000 -08:00 a11a89f8
│  added line b to file1
◉  kppvnyqs john@gilfoyle.local 2023-12-11 08:51:29.000 -08:00 78a03ac2
│  added file1 with line a
◉  zzzzzzzz root() 00000000
```
Again, confused here. I don’t understand the branch - pr - merge - history conceptual model. At the end of the day, I was hoping for a clean commit log without branches. 

And, I even like to push my working copy assuming that’s ok so I can pull it down from another machine, but when I do that the history confused me. I would expect to see that as an @ at the top of the log with linear commit history.

### Confusion

- The history appears non-linear post-merge and fetch.
- Unsure about the mapping between the JJ history and GitHub repo state.
- Expectation was a linear history throughout the process.
- 

### Request for Clarification

I understand this might be a user error, but I'm struggling to comprehend the discrepancy between my expectations and the actual JJ log outputs. Any insights or explanations would be greatly appreciated.

---
*Originally posted on Hacker News discussion [here](https://news.ycombinator.com/item?id=38590080#38593087)*

**Comment by martinvonz** [comment]
> At this point I become confused. I expect the history to be linear, but I see what appears to be a branch.

I think it's GitHub that created the merge commit. The default way of "merging" a PR is to do a `git merge --no-ff`. See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges. Does that resolve your confusion?

**Reply by ilyagr** [reply, marked as answer]
I think @martinvonz described what happened at **step 6** of the question. I'll try to explain what happened at step 5.

At **step 5**, the commit `vvlz` that looks like a branch is the so-called "working-copy commit", marked by a `@` symbol. `jj` created that commit, but note that it is marked as "empty", meaning it has no diffs from its parent commit. So, the "branch" is trivial at this point. If you edit the working copy, the working copy commit is the one that will be modified. At that point, you'd have a real branch, since you are adding changes on top of commit `krzmz`. If you wanted to get rid of the branch, you could amend the changes you made into `krzmz`, but `jj` doesn't do it until you explicitly tell it to.

OTOH, if you don't edit the working copy, you can pretty much ignore this commit `vvlz`. For example, if you check out another commit (e.g. `jj co rokr`) while `vvlz` one is still empty, commit `vvlz` will disappear, and **another empty commit** will appear on top of `rokr`.

I'm not entirely sure why, after `jj git clone`, `jj` made the working copy be on top of `krzmz` as opposed to `rokr`. I'm guessing that has to do with what GitHub thought was the default branch.

**Reply by BatmanAoD** [reply]
I think you're spot-on that GitHub made `push-zlpppmmsqsyu` the default branch, since that was the first branch to exist.

**Reply by johnwheeler** [reply]
> I think it's GitHub that created the merge commit

What should I have github do?

**Reply by martinvonz** [reply]
It's up to you if it's your own project :) I personally prefer to rebase, but lots of people prefer to merge. I'm just saying that if you use the merge setting, then it's expected to get merge commits when you merge the PR (GitHub calls it "merging" a PR even if you have the setting set to merge).
